### PR TITLE
SCRS-11442 - ALF design update

### PR DESCRIPTION
**Pull request by ChrisHillCap**
Updated as per story (please see story SCRS-11442 for full details of changes)
notes:
- Postcode is optional on edit pages, if user enters postcode and country is UK, postcode format is validated
- In some situations, when retrieving an address from ALF via the API, the county will be the final thing at the end of the "address lines" instead of town (to mimic the order of uk edit mode page). this should not affect anyone unless they were relying on the final entry in the list in the jsobject to be town.
- added in IT tests
- cleared "selected" address on lookup page load (GET) and added redirect on confirm page if selected address does not exist, which redirects back to lookup page
- we noticed that the "limit" for the results on the select page does NOT have a default value so have changed the readme accordingly
- added test route /lookup-address/test-only/test-setup which allows for user to enter own json setup block without interacting with another service, upon clicking continue on confirm page user will see json stored in keystore including confirmed address an custom config etc

### CODE REVIEW

**Comment by ChrisHillCap** [comment]
There is a service manager config change for this which also has a PR, but I believe has not been merged, needs to be merged after merging this

File: app/utils/IsCountryUK.scala [hunk 1]
```diff
@@ -0,0 +1,12 @@
+package utils
+
+import model.ConfirmableAddress
+
+object IsCountryUK {
+  val countryCheck: Option[ConfirmableAddress] => Boolean = (selectedAddress: Option[ConfirmableAddress]) => {
+    selectedAddress.fold(false)(
+      _.address.country.fold(true)(_.code == "GB")
+    )
```
Comment (hughfdjackson) [inline app/utils/IsCountryUK.scala]: Same comment here as above - I find the above hard to follow.  The logic itself is a little tricky to express (since not all Nones get defaulted to the same value), but you might get further with splitting it up into separate functions?

Maybe something like:

```scala
def isConfirmableAddressInUk(confirmableAddress: ConfirmableAddress) = 
  confirmableAddress.address.country
    .map(_.code == "GB")
    .getOrElse(true)

selectedAddress.map(isConfirmableAddressInUK).getOrElse(false)
```

File: app/utils/PostcodeHelper.scala [hunk 2]
```diff
@@ -0,0 +1,9 @@
+package utils
+
+import uk.gov.hmrc.address.uk.Postcode
+
+object PostcodeHelper {
+
+  def displayPostcode(p:Option[String]) = p.fold("")(pc => Postcode.cleanupPostcode(pc).fold("")(_.toString))
+  def displayPostcode(p:String) = Postcode.cleanupPostcode(p).fold("")(_.toString)
```
Comment (hughfdjackson) [inline app/utils/PostcodeHelper.scala]: :D this is become a bit of an early-morning code-golf. 

```scala
   def displayPostcode(p:String) = Postcode.cleanupPostcode(p).map(_.toString).getOrElse("")
```

File: app/utils/PostcodeHelper.scala [hunk 3]
```diff
@@ -0,0 +1,9 @@
+package utils
+
+import uk.gov.hmrc.address.uk.Postcode
+
+object PostcodeHelper {
+
+  def displayPostcode(p:Option[String]) = p.fold("")(pc => Postcode.cleanupPostcode(pc).fold("")(_.toString))
```
Comment (hughfdjackson) [inline app/utils/PostcodeHelper.scala]: I didn't realise until this morning how many different patterns you could get out of fold!  

Would the equivalent be something like this? 
```scala
   def displayPostcode(p:Option[String]) = p.flatMap(Postcode.cleanupPostcode).map(_.toString).getOrElse("")
```

File: it/controllers/AddressLookupControllerISpec.scala [hunk 4]
```diff
@@ -0,0 +1,297 @@
...
+      stubKeystore("Jid123", Json.toJson(configWithConfirmedAddress).as[JsObject], 200)
+      val fResponse = buildClient("confirmed?id=Jid123", journeyID = "", lookUpOrApi = "api").
+        withHeaders(HeaderNames.COOKIE -> sessionCookie, "Csrf-Token" -> "nocheck").
+        get()
+      val res = await(fResponse)
+      res.status shouldBe 200
+      res.json shouldBe Json.toJson(ConfirmableAddress("foo",Some("bar"),ConfirmableAddressDetails(Some(List("1","","3","4")),Some("wizz"),None)))
+    }
+  }
+}
```
[the remark this answers is not in the document]
Comment (ChrisHillCap) [reply it/controllers/AddressLookupControllerISpec.scala]: 👍

File: test/forms/ALFFormsSpec.scala [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+import model.Edit
+import org.scalatest.{MustMatchers, WordSpec}
+import play.api.data.validation.{Invalid, Valid}
+
+class ALFFormsSpec extends WordSpec with MustMatchers{
+
+  val editFormNonuk = ALFForms.nonUkEditForm
+  val editFormUk = ALFForms.ukEditForm
+
+  val chars257 = List.fill(257)("A").reduce(_ + _)
```
[the remark this answers is not in the document]
Comment (ChrisHillCap) [reply test/forms/ALFFormsSpec.scala]: ok i wont change then

File: app/controllers/testonly/StubController.scala [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+
+class StubControllerImpl @Inject()(val apiController: ApiController,val  journeyRepository: JourneyRepository)(implicit val ec: ExecutionContext, implicit val messagesApi: MessagesApi) extends StubController {
+
+}
+object TestSetupForm {
+  val form = Form(single(
+    "journeyConfig" -> text
+  ))
+}
+@ImplementedBy(classOf[StubControllerImpl])
```
[the remark this answers is not in the document]
Comment (hughfdjackson) [reply app/controllers/testonly/StubController.scala]: A review comment - if there wasn't a specific reason to separate implementation and interface definition, I'd prefer it to be a single class.

File: app/model/Model.scala [hunk 7]
```diff
@@ -17,22 +17,15 @@ case class Select(addressId: String)
 
 case class Edit(line1: String, line2: Option[String], line3: Option[String], town: String, postcode: String, countryCode: Option[String]) {
 
-  def toConfirmableAddress(auditRef: String, isukMode: Boolean): ConfirmableAddress = ConfirmableAddress(
+    def toConfirmableAddressUkAndNonUk(auditRef: String): ConfirmableAddress = ConfirmableAddress(
```
[the remark this answers is not in the document]
Comment (ChrisHillCap) [reply app/model/Model.scala]: ok

File: app/utils/IsCountryUK.scala (hunk 1 above)
Comment (ChrisHillCap) [reply app/utils/IsCountryUK.scala]: again i think this is a per team basis.. our team is fine with folds.. I will change

File: app/utils/PostcodeHelper.scala (hunk 3 above)
Comment (ChrisHillCap) [reply app/utils/PostcodeHelper.scala]: done

File: app/utils/PostcodeHelper.scala (hunk 2 above)
Comment (ChrisHillCap) [reply app/utils/PostcodeHelper.scala]: done

File: app/utils/IsCountryUK.scala (hunk 1 above)
Comment (ChrisHillCap) [reply app/utils/IsCountryUK.scala]: actually i disagree, i think it is readable <.<

Comment (hughfdjackson) [reply app/utils/IsCountryUK.scala]: Definitely a subjective thing - I would appreciate the change in any case, though.  Of the two of us, at least one of us finds the fold-y version a bit difficult to parse. 

Happy to let @thomasrynnehmrc be the tie-breaker if you feel strongly against a change similar to the suggestion.

File: app/controllers/testonly/StubController.scala (hunk 6 above)
[the remark this answers is not in the document]
Comment (ChrisHillCap) [reply app/controllers/testonly/StubController.scala]: done

File: app/utils/IsCountryUK.scala (hunk 1 above)
Comment (thomasrynnehmrc) [reply app/utils/IsCountryUK.scala]: I find the folds hard to read. Actually In this case I think I would use match :) but I prefer map getOrElse over fold.

**Comment by ChrisHillCap** [comment]
Here are your options: please pick one           val countryCheck: Option[ConfirmableAddress] => Boolean = (selectedAddress: Option[ConfirmableAddress]) => {
    selectedAddress
      .map(
        _.address.country.map(_.code == "GB").getOrElse(true)
      ).getOrElse(false)
  }

  def countryCheck(selectedAddress: Option[ConfirmableAddress]): Boolean = {
   def isConfirmableAddressInUk(confirmableAddress: ConfirmableAddress): Boolean = { confirmableAddress.address.country
    .map (_.code == "GB")
    .getOrElse (true)
   }
    selectedAddress.map(isConfirmableAddressInUk).getOrElse(false)
  }
